metrics/opentelemetry.go

```diff
@@ -0,0 +1,226 @@
+	if err != nil {
+		return nil, err
+	}
+
+	extractorRetries, err := meter.Int64Counter("meteor.extractor.retries")
+	if err != nil {
+		return nil, err
+	}
+
+	assetsExtracted, err := meter.Int64Counter("meteor.assets.extracted")
+	if err != nil {
+		return nil, err
+	}
+
+	sinkRetries, err := meter.Int64Counter("meteor.sink.retries")
+	if err != nil {
+		return nil, err
+	}
+
+	return &OtelMonitor{
+		recipeDuration:   recipeDuration,
+		extractorRetries: extractorRetries,
+		assetsExtracted:  assetsExtracted,
+		sinkRetries:      sinkRetries,
+	}, nil
+}
+
+func initGlobalMetrics(ctx context.Context, res *resource.Resource, cfg config.Config, logger *log.Logrus) (func(), error) {
+	exporter, err := otlpmetricgrpc.New(ctx,
+		otlpmetricgrpc.WithEndpoint(cfg.OtelCollectorAddr),
+		otlpmetricgrpc.WithCompressor(gzip.Name),
+		otlpmetricgrpc.WithInsecure(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("create metric exporter: %w", err)
+	}
+
+	reader := sdkmetric.NewPeriodicReader(
+		exporter,
+		sdkmetric.WithInterval(15*time.Second),
+	)
+
+	provider := sdkmetric.NewMeterProvider(
+		sdkmetric.WithReader(reader),
+		sdkmetric.WithResource(res),
+	)
+	otel.SetMeterProvider(provider)
+
+	return func() {
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), gracePeriod)
+		defer cancel()
+		if err := provider.Shutdown(shutdownCtx); err != nil {
+			logger.Error("otlp metric-provider failed to shutdown", "err", err)
+		}
+	}, nil
+}
+
+func initGlobalTracer(ctx context.Context, res *resource.Resource, cfg config.Config, logger *log.Logrus) (func(), error) {
+	exporter, err := otlptrace.New(ctx, otlptracegrpc.NewClient(
+		otlptracegrpc.WithEndpoint(cfg.OtelCollectorAddr),
+		otlptracegrpc.WithInsecure(),
+		otlptracegrpc.WithCompressor(gzip.Name),
+	))
+	if err != nil {
+		return nil, fmt.Errorf("create trace exporter: %w", err)
+	}
+
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(consistent.ProbabilityBased(cfg.OtelTraceSampleProbability)),
+		sdktrace.WithResource(res),
+		sdktrace.WithSpanProcessor(sdktrace.NewBatchSpanProcessor(exporter)),
+	)
+
+	otel.SetTracerProvider(tracerProvider)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propagation.Baggage{},
+	))
+
+	return func() {
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), gracePeriod)
+		defer cancel()
+		if err := tracerProvider.Shutdown(shutdownCtx); err != nil {
+			logger.Error("otlp trace-provider failed to shutdown", "err", err)
+		}
+	}, nil
+}
+
+func getSliceStringPluginNames(prs []recipe.PluginRecipe) []string {
+	var res []string
+	for _, pr := range prs {
+		res = append(res, pr.Name)
+	}
+
+	return res
+}
+
+// RecordRun records a run behavior
+func (m *OtelMonitor) RecordRun(ctx context.Context, run agent.Run) {
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(attribute.String("recipe_name", run.Recipe.Name))
+
```

Comment:
```suggestion
```

Don't think this would be useful since we are not instrumenting traces.

Reply:
deleted